sw/n64/testrom/shell.c

```diff
@@ -0,0 +1,188 @@
+
+/*
+ * Render a list of strings and show a caret on the currently selected row
+ */
+void render_list(display_context_t display, char *list[], int currently_selected, int first_visible, int max_on_screen) {
+
+    /* If we aren't starting at the first row, draw an indicator to show more files above*/
+    if (first_visible > 0) {
+        graphics_draw_text(display, MARGIN_PADDING, MENU_BAR_HEIGHT, "...");
+    }
+
+    for (int i = 0; i < max_on_screen; i++) {
+        int row = first_visible+i;
+        int x = currently_selected == row ? MARGIN_PADDING : 0; // if the current item is selected move it over so we can show a selection caret
+        int y = i * ROW_HEIGHT + LIST_TOP_PADDING; // First row must start below the menu bar, so add that plus a pad
+        
+        // if we run out of items to draw on the screen, don't draw anything else
+        if (row > NUM_ENTRIES) {
+            break;
+        }
+
+        if (currently_selected == row) {
+            /* Render selection box */
+            graphics_draw_box(display, 0, y-2, ROW_SELECTION_WIDTH, 10, graphics_convert_color(SELECTION_COLOR));
+
+            /* Now render selection caret */
+            graphics_draw_text(display, MARGIN_PADDING/2, y, "  >");
+
+            /* Increase Margin Padding to account for added caret width */
+            x += MARGIN_PADDING;
+        } else {
+            x += MARGIN_PADDING;
+        }
+
+        // Render the list item
+        graphics_draw_text(display, x, y, list[row]);
+
+        /* If this is the last row and there are more files below, draw an indicator*/
+        if (row+1 < NUM_ENTRIES && i + 1 >= max_on_screen) {
+            graphics_draw_text(display, MARGIN_PADDING, y + 10, "...");
+        }
+    }
+}
+
+void draw_header_bar(display_context_t display, int fileCount) {
+    int x = 0, y = 0, width = SCREEN_WIDTH, height = MENU_BAR_HEIGHT;
+    graphics_draw_box(display, x, y, width, height, graphics_convert_color(MENU_BAR_COLOR));
+    
+    char menuHeaderBuffer[50];
+    sprintf(menuHeaderBuffer, "PicoCart64 OS b2022.07.17\t\t\t\t%d Files\0", fileCount);
+    graphics_draw_text(display, MARGIN_PADDING, y+4, menuHeaderBuffer);
+}
+
+/*
+ * Init display and controller input then render a list of strings, showing currently selected
+ */
+void show_list() {
+    // For now, hard code some strings in so we can have something to test with
+    char *entries[21] = {
+        "Goldeneye 007 (USA).n64", 
+        "Mario 64 (USA).n64", 
+        "Star Wars Rouge Squadron (USA).n64", 
+        "Star Wars Shadows of the Empire (USA).n64", 
+        "The Legend of Zelda, Ocarina of Time (USA).n64", 
+        "Perfect Dark (USA).n64", 
+        "Mario Kart 64 (USA).n64", 
+        "Killer Instinct 64 (USA).n64", 
+        "Star Wars Podracer (USA).n64", 
+        "007 Tomorrow Never Dies (USA).n64",
+        "Donkey Kong 64 (USA).n64",
+        "testrom.n64",
+        /*"お母さん３(JPN).n64",*/ // I don't think it likes this entry too much
```

Comment:
We would probably need to add a font for this. Haven't looked too much at how libdragon renders text, but we would probably want to keep strings in UTF-8 as well.